app/features/receive/use-receive-cashu-token.tsx

```diff
@@ -0,0 +1,349 @@
+) => {
+  const badges: string[] = [];
+
+  if (sourceAccount.isTestMint) {
+    badges.push('Test Mint');
+  }
+  if (account.mintUrl === sourceAccount.mintUrl) {
+    badges.push('Source');
+  }
+  if (
+    defaultAccount.type === 'cashu' &&
+    account.mintUrl === defaultAccount.mintUrl
+  ) {
+    badges.push('Default');
+  }
+  if (!allAccounts.some((a) => a.mintUrl === account.mintUrl)) {
+    badges.push('Unknown');
+  }
+
+  return badges;
+};
+/**
+ * Hook to receive a cashu token.
+ * @param token - The token to receive
+ * @param cashuPubKey - The user's cashu public key
+ * @returns Token data fetched by the query function and functions to claim the token.
+ * Data is undefined while isLoading is true.
+ */
+export function useReceiveCashuToken({
+  token,
+  cashuPubKey,
+}: UseReceiveCashuTokenProps): UseReceiveCashuTokenReturn {
+  const [isClaiming, setIsClaiming] = useState(false);
+  const { toast } = useToast();
+  const navigate = useNavigateWithViewTransition();
+  const [selectedReceiveAccount, setSelectedReceiveAccount] = useState<
+    CashuAccount | undefined
+  >();
+  const { data: allAccounts } = useAccounts();
+  const accounts = allAccounts?.filter(
+    (a) => a.type === 'cashu',
+  ) as CashuAccount[];
+  const addAccount = useAddAccount();
+  const defaultAccount = useDefaultAccount();
+
+  const { data: sourceAccountData, isLoading: isSourceAccountLoading } =
+    useQuery({
+      queryKey: ['mint-info', token.mint],
+      queryFn: async () => {
+        const existingAccount = accounts.find((a) => a.mintUrl === token.mint);
+        if (existingAccount) {
+          return existingAccount;
+        }
+
+        const info = await getMintInfo(token.mint);
+        const isTestMint = await checkIsTestMint(token.mint);
+        return tokenToSourceAccount(token, info, isTestMint);
+      },
+    });
+
+  const { data: tokenData, isLoading: isTokenDataLoading } = useQuery({
+    queryKey: ['token-state', token, sourceAccountData],
+    enabled: !!sourceAccountData,
+    queryFn: async () => {
+      if (!sourceAccountData) {
+        return undefined;
+      }
+
+      const unspentProofs = await getUnspentProofsFromToken(token);
+      const { claimableProofs, cannotClaimReason: initialReason } =
+        getClaimableProofs(unspentProofs, cashuPubKey);
+
+      let cannotClaimReason = initialReason;
+      const isSpent = unspentProofs.length === 0;
+      if (isSpent) {
+        cannotClaimReason = 'This ecash has already been spent';
+      } else if (claimableProofs.length === 0) {
+        cannotClaimReason = 'You do not have permission to claim this ecash';
+      }
+
+      const canClaim = claimableProofs.length > 0;
+      const displayToken: Token = canClaim
+        ? { ...token, proofs: claimableProofs }
+        : token;
+      const money = tokenToMoney(displayToken);
+      const isMintKnown = accounts.some(
+        (a) => a.type === 'cashu' && a.mintUrl === token.mint,
```

Comment:
do we even need to check the type here if we already filtered for cashu accounts?